File: swift/lib/Sources/Client/RemoteClient.swift
public class RemoteClient {
    internal class RemoteClientReturn<T: Decodable>: Decodable {
        var r: T?
    }

    public static func call<T: Decodable>(_ request: Request, _ callback: ((_ result: T?) -> Void)? = nil) {
        let key = UniqueID.generate()

        CallbackList.shared.add(key: key) { (response: String) in
            guard let data: RemoteClientReturn<T> = XPLPC.shared.config.serializer.decodeFunctionReturnValue(response) else {
                Log.e("[RemoteClient : call] Error when decode data")
                callback?(nil)
                return
            }

            callback?(data.r)
        }

        PlatformProxy.shared.nativeProxyCall(key, request.data)
    }
}

File: kotlin/lib/library/src/main/java/com/xplpc/client/RemoteClient.kt
package com.xplpc.client

import com.google.gson.reflect.TypeToken
import com.xplpc.core.XPLPC
import com.xplpc.data.CallbackList
import com.xplpc.message.Request
import com.xplpc.proxy.PlatformProxy
import com.xplpc.util.Log
import com.xplpc.util.UniqueID

object RemoteClient {
    inline fun <reified T> call(request: Request, noinline callback: ((T?) -> Unit)?) {
        val key = UniqueID.generate()

        CallbackList.add(key) { response ->
            val type = object : TypeToken<T>() {}

            val data: T? = try {
                XPLPC.config.serializer.decodeFunctionReturnValue(response, type)
            } catch (e: Exception) {
                Log.e("[RemoteClient : call] Error when decode data: ${e.message}")
                null
            }

            callback?.invoke(data)
        }

        PlatformProxy.nativeProxyCall(key, request.data)
    }

    inline fun <reified T> call(request: Request) {
        call<T>(request, null)
    }
}
